cpp/include/cuml/experimental/fil/detail/infer_kernel/cpu.hpp

```diff
@@ -117,48 +118,74 @@ void infer_kernel_cpu(
         auto tree_output = std::conditional_t<
           has_vector_leaves, typename node_t::index_type, typename node_t::threshold_type
         >{};
-        if constexpr (has_nonlocal_categories) {
-          tree_output = evaluate_tree<has_vector_leaves>(
-            forest.get_tree_root(tree_index),
-            input + row_index * col_count,
-            categorical_data
-          );
+        auto leaf_node_id = index_type{};
```

Comment:
I'm not entirely sure that this whole section needs to be quite as complex as it is. Part of the point of letting the type of the id mapping pointer dictate the inference type is that we should not need specialized logic for different inference types at this depth of the code. Given that the indexing logic can already be a little tricky here, is there a way to simplify this?

More specifically, can you point to a reason why we need to specifically reference the inference type outside of the sort of tweaks to indexing that we did with `predict_per_tree`?

Reply:
> can you point to a reason why we need to specifically reference the inference type outside of the sort of tweaks to indexing that we did with predict_per_tree?

This is because `apply()` should not use a `for` loop even when `has_vector_leaves` is true.

Reply:
> How should we conditionally pass parameters forest.get_tree_root(0) and forest.get_node_id_mapping()?

Maybe we shouldn't pass them conditionally. It's relatively cheap to fetch the first tree root, so it would probably be fine to do that for ordinary prediction. As for the mapping, we do need to differentiate the inference type to call the correct `evaluate_tree` instantiation, but maybe we can do that in a more self-contained section of code, either through wrapping it in a well-named function or by simplifying the branching that's going on here.

> This is because apply() should not use a for loop even when has_vector_leaves is true.

Aha! Yep, that's true. We should be able to just add the inference type check to the has_vector_leaves check in that case.

Reply:
I created a separate function to handle the dispatching to the correct `evaluate_tree` kernel.